824 Revise Security &  Privacy chapter

profvjreddi opened the pull request:
https://github.com/harvard-edge/cs249r_book/issues/824

18jeffreyma commented:
- [ ] 15.3.2: System-Level Isolation: I feel like some diagram could be used here to explain how the isolation should work (i.e. architecturally blocking and compartmentalizing (something like https://help.apple.com/assets/663BFE394E05E4CA5D0A02DC/663BFE3C4E05E4CA5D0A02E8/en_US/9302f62027e1966222080ef2731e45de.png)

- [ ] 15.4.1 should mention https://arxiv.org/abs/2403.06634 (stealing OpenAI architecture, very recent work and in ML security space), i would put this under approximate model theft

- [ ] 15.4.4 maybe a picture of this adversarial sticker from the paper to hammer in this point?
- [ ] 15.5.2 diagram?
- [ ] Figure 15.5 could use some more annotation with the jump that skips the incorrect behavior?
- [ ] 15.5.4 could pull some diagrams from http://hardwarefail.ml/
- [ ] 15.5.6: i wonder if theres any case studies here on purchasing conterfeit hardware (or bring in a quick reference before diving into the case study in 15.5.8)
- [ ] 15.5.8 needs a reference to the super micro controversy news link i think (just to cite sources and allow readers to dive deeper)
- [ ] Output Monitoring: I think maybe some content from the nibbler paper would be good here to explain why prompt filtering is important + why its such a hard a tricky problem?
- [ ] Additionally, some references to recent work like https://arxiv.org/html/2312.06674v1 or ShieldGemma https://ai.google.dev/gemma/docs/shieldgemma

18jeffreyma commented:
i think other vibes based feedback: i think this section defintely makes sense from a security/privacy point of view, but i wonder if theres additional ML security + privacy specific stuff.

i liked the callouts to differential privacy and model stealing as well as hardware fault injection attacks. i think we could def add some of the ai safety stuff (i.e. judging safe model outputs) as well as I think a really good add is MCP vulnerability (which aligns with stuff like prompt safety as now we're trying to hook up LLMs with untrusted servers, which is inviting a whole new ML specific security problem subspace)

https://www.upwind.io/feed/unpacking-the-security-risks-of-model-context-protocol-mcp-servers

profvjreddi commented:
> i think other vibes based feedback: i think this section defintely makes sense from a security/privacy point of view, but i wonder if theres additional ML security + privacy specific stuff.

Yeah, that's one thing I also felt we are missing. 

> i liked the callouts to differential privacy and model stealing as well as hardware fault injection attacks. i think we could def add some of the ai safety stuff (i.e. judging safe model outputs) as well as I think a really good add is MCP vulnerability (which aligns with stuff like prompt safety as now we're trying to hook up LLMs with untrusted servers, which is inviting a whole new ML specific security problem subspace)
> 
> https://www.upwind.io/feed/unpacking-the-security-risks-of-model-context-protocol-mcp-servers

Good points. I have been wondering how to best make a pass through the book for integrating aspects related to LLMs, cause they stick out if I try to cram them in. But I am sure from this chapter pov we can do that.

profvjreddi commented:
> * [x]  15.3.2: System-Level Isolation: I feel like some diagram could be used here to explain how the isolation should work (i.e. architecturally blocking and compartmentalizing (something like https://help.apple.com/assets/663BFE394E05E4CA5D0A02DC/663BFE3C4E05E4CA5D0A02E8/en_US/9302f62027e1966222080ef2731e45de.png)

Good point. I do this in the TEE section, but based on your feedback I have added some more details and explained how the FaceID on the apple iPhone works using the secure enclave methods.

> * [x]  15.4.1 should mention https://arxiv.org/abs/2403.06634 (stealing OpenAI architecture, very recent work and in ML security space), i would put this under approximate model theft

Woo... good one! Thanks! I merged this into the model threat section.

> * [ ]  15.4.4 maybe a picture of this adversarial sticker from the paper to hammer in this point?

👍  Done!

> * [ ]  15.5.2 diagram?

👍  Done! 

> * [ ]  Figure 15.5 could use some more annotation with the jump that skips the incorrect behavior?

👍  Done!

> * [ ]  15.5.4 could pull some diagrams from http://hardwarefail.ml/

This is a good one, I will put this stuff in the Robust AI chapter.

> * [ ]  15.5.6: i wonder if theres any case studies here on purchasing conterfeit hardware (or bring in a quick reference before diving into the case study in 15.5.8)

mm... could not find any :( 

> * [ ]  15.5.8 needs a reference to the super micro controversy news link i think (just to cite sources and allow readers to dive deeper)

Good catch, added it!

> * [ ]  Output Monitoring: I think maybe some content from the nibbler paper would be good here to explain why prompt filtering is important + why its such a hard a tricky problem?

Done!

> * [ ]  Additionally, some references to recent work like https://arxiv.org/html/2312.06674v1 or ShieldGemma https://ai.google.dev/gemma/docs/shieldgemma

🙏 

Thanks for all the good feedback -- incorporated it all!
